Localize Enum Display names

ryanbrandenburg opened the pull request:
This PR addresses a bug and a couple missing features.

The bug (#5198) is that Enum display values  would previously store the first value that was used and always use that. By passing a function which returns a Name value to the constructor of EnumGroupAndName and executing that when we need the name we avoid this.

One missing feature is #5197, that `GetEnumSelectList` was checking the DisplayAttributes for values to display, but it wasn't localizing those values like other uses of DisplayAttributes do.

The other missing feature was #4215. It boils down to that `Html.DisplayFor` didn't pay attention to DisplayAttributes on enums, and consequently didn't localize them.

ryanbrandenburg commented on src/Microsoft.AspNetCore.Mvc.Abstractions/ModelBinding/EnumGroupAndName.cs:
```diff
@@ -28,7 +30,23 @@ public EnumGroupAndName(string group, string name)
+            _nameFunc = () => name;
+        }
+
+        public EnumGroupAndName(string group, Func<string> name)
```
I couldn't decide if we should `Obsolete` the old constructor. My instinct was no, to allow people to do the simple thing if they didn't need anything complex, but I fear that will need to ignoring the new constructor and getting us right back into the same spot.

Also, going back to add comments to this an any new public methods now.

Eilon replied:
Yeah I'm not sure if it needs to be obsoleted. If the docs just said precisely what it did and didn't do, that would seem to be fine.

dougbu commented:
I'll take reviewing this one.

ryanbrandenburg commented:
CC @Eilon and @dougbu. I know this isn't exactly what we had talked about (adding new Enum properties to the DisplayName context) but once I got down into it it seemed cleaner this way.

dougbu commented:
> The bug is that Enum value localization

When submitting the final commit, please update the description to avoid the implication that `[Display(Name = "Fred")]` is only about localization. The attribute is in general about UI overrides, in some cases including localization of the displayed strings.

dougbu commented on src/Microsoft.AspNetCore.Mvc.Abstractions/ModelBinding/EnumGroupAndName.cs:
```diff
@@ -28,7 +30,28 @@ public EnumGroupAndName(string group, string name)
+        /// </summary>
+        /// <param name="group">The group name.</param>
+        /// <param name="name">A function which returns the name. (Necessary for proper localization.)</param>
+        public EnumGroupAndName(string group, Func<string> name)
```
Probably should `[Obsolete]` the old constructor and file a 2.0 bug about removing it.

dougbu commented on src/Microsoft.AspNetCore.Mvc.Abstractions/ModelBinding/EnumGroupAndName.cs:
```diff
@@ -28,7 +30,28 @@ public EnumGroupAndName(string group, string name)
 
             Group = group;
-            Name = name;
+            _nameFunc = () => name;
```
Suggest shortening this constructor to
``` c#
public EnumGroupAndName(string group, string name)
    : this(group, () => name)
{
}
```

ryanbrandenburg replied:
That would change the behavior because passing a null `name` would turn into a function that returns null and then not throw, right? It's a small difference, but it doesn't really seem worth it to me for some internal cleanup.

dougbu commented on src/Microsoft.AspNetCore.Mvc.DataAnnotations/Internal/DataAnnotationsMetadataProvider.cs:
```diff
@@ -153,15 +153,15 @@ public void CreateDisplayMetadata(DisplayMetadataProviderContext context)
 
                     groupedDisplayNamesAndValues.Add(new KeyValuePair<EnumGroupAndName, string>(
-                        new EnumGroupAndName(groupName, displayName),
+                        new EnumGroupAndName(groupName, () => GetDisplayName(enumLocalizer, field)),
```
Check whether this capturing lambda chews through more memory than instantiating the `EnumGroupAndName` with the `IStringLocalizer` and `FieldInfo`. Don't need `GetDisplayName()` in this class if that works better.

dougbu commented on src/Microsoft.AspNetCore.Mvc.DataAnnotations/Internal/DataAnnotationsMetadataProvider.cs:
```diff
@@ -153,15 +153,15 @@ public void CreateDisplayMetadata(DisplayMetadataProviderContext context)
                 // Dictionary does not guarantee order will be preserved.
                 var groupedDisplayNamesAndValues = new List<KeyValuePair<EnumGroupAndName, string>>();
                 var namesAndValues = new Dictionary<string, string>();
+                IStringLocalizer enumLocalizer = _stringLocalizerFactory?.Create(underlyingType);
```
`var`

dougbu commented on src/Microsoft.AspNetCore.Mvc.DataAnnotations/Internal/DataAnnotationsMetadataProvider.cs:
```diff
@@ -301,6 +301,10 @@ private static string GetDisplayName(FieldInfo field)
                 var name = display.GetName();
                 if (name != null)
                 {
+                    if (localizer != null && name != string.Empty && display.ResourceType == null)
```
`!string.IsNullOrEmpty(name)`

dougbu commented on src/Microsoft.AspNetCore.Mvc.ViewFeatures/Internal/ExpressionMetadataProvider.cs:
```diff
@@ -104,6 +93,34 @@ public static class ExpressionMetadataProvider
+                try
+                {
+                    object result = CachedExpressionCompiler.Process(expression)((TModel)container);
+                    if (expression.ReturnType.GetTypeInfo().IsEnum)
```
No need for this check to be inside the lambda -- `expression` means it's already capturing. So calculate `var isEnum` once and use it here.

dougbu replied:
Actually, better to not recalculate this at all. Instead use `metadata.EnumNamesAndValues != null`.

dougbu commented on src/Microsoft.AspNetCore.Mvc.ViewFeatures/Internal/ExpressionMetadataProvider.cs:
```diff
@@ -104,6 +93,34 @@ public static class ExpressionMetadataProvider
+                    object result = CachedExpressionCompiler.Process(expression)((TModel)container);
+                    if (expression.ReturnType.GetTypeInfo().IsEnum)
+                    {
+                        var enumValue = metadata.EnumNamesAndValues[result.ToString()];
```
Won't `enumValue == result` be `true` in all cases except when the `enum` has a value that doesn't match one of its fields e.g. `(StringComparison)23`?

ryanbrandenburg replied:
I don't think I follow what you mean by
>enum has a value that doesn't match one of its fields

Could you maybe rephrase that or give an expanded example?

dougbu replied:
1. I don't get what this line is for.
2. The indexer will blow up if `metadata.ModelType == typeof(StringComparison)` and `result` isn't between `CurrentCulture` (`0`) and `OrdinalIgnoreCase` (`5`) e.g. `(StringComparison)23`. That's completely valid but unsupported here.
3. Further `result` can easily be `null` when `metadata.ModelType` is a `Nullable<TEnum>`. The indexer will blow up then too.

dougbu commented on src/Microsoft.AspNetCore.Mvc.ViewFeatures/Internal/ExpressionMetadataProvider.cs:
```diff
@@ -104,6 +93,34 @@ public static class ExpressionMetadataProvider
+                        var enumValue = metadata.EnumNamesAndValues[result.ToString()];
+                        foreach (var kvp in metadata.EnumGroupedDisplayNamesAndValues)
+                        {
+                            if (kvp.Value == enumValue)
```
Always use `string.Equals(...)` when doing `string` comparisons. Also be explicit about the `StringComparison`.

dougbu commented on src/Microsoft.AspNetCore.Mvc.ViewFeatures/Internal/ExpressionMetadataProvider.cs:
```diff
@@ -104,6 +93,34 @@ public static class ExpressionMetadataProvider
                 metadata = metadataProvider.GetMetadataForType(containerType).Properties[propertyName];
             }
 
+            Func<object, object> modelAccessor = (container) =>
```
I'm very hesitant about making any changes to this `Func`. It's just too general. For example, your changes here will affect how `enum`s are displayed in `<input/>` tag helpers. Think we want something specific to `@Html.Display[For]()`, `@Html.DropDownList[For]()`, `@Html.ListBox[For]()`, and the `<select>` tag helper.

ryanbrandenburg replied:
Why wouldn't we want the display string of enums to be the same across all of our elements and helpers?

dougbu replied:
... because the other locations are about values and not the names of values. Think of the difference between the `Text` and `Value` properties of a `SelectListItem` or (equivalently) between the body of an `<option>` element and its `value` attribute.

Bottom line, don't make changes beyond the scope of the identified problem -- bodies of generated `<option>` elements and the `@Html.Display[For]()` return value.

Eilon replied:
I agree with @dougbu that this *must not* affect the behavior of `<input>` tags because those are about the *value* of the enum, not the *display text* of the enum.

dougbu commented on test/Microsoft.AspNetCore.Mvc.ViewFeatures.Test/Rendering/HtmlHelperDisplayExtensionsTest.cs:
```diff
@@ -355,6 +414,22 @@ public void DisplayForModel_UsesTemplateNameAndHtmlFieldName()
+        {
+            [Display(Name = "Resx_Pass")]
+            Pass,
+            [Display(Name = "Type_Fail", ResourceType = typeof(StatusResource))]
```
This is in no way a failure. Just a different way to override the display name.

Need to cover `[Display(ResourceType = typeof(StatusResource))]` here.

Separately, please confirm the four cases for these files (after the above addition) are covered when the string localizer factory is `null`.

ryanbrandenburg replied:
I'm not suggesting that it's a failure, Fail is the name of the enum value it's representing.

dougbu replied:
Let's do something less confusing like `Created`, `Started`, `Faulted`, and `DoneDoneDone`.

dougbu commented on test/Microsoft.AspNetCore.Mvc.DataAnnotations.Test/Internal/DataAnnotationsMetadataProviderTest.cs:
```diff
@@ -727,6 +772,91 @@ public void CreateDisplayMetadata_IsFlagsEnum_ReflectsModelType(Type type, bool
+            var stringLocalizer = new Mock<IStringLocalizer>(MockBehavior.Strict);
+            stringLocalizer
+                .Setup(loc => loc["LOC_Two"])
+                .Returns<string>((k => new LocalizedString(k, $"{k} {CultureInfo.CurrentCulture}")));
```
When mocking things this way, why is `DataAnnotationsMetadataProviderTest.EnumWithLocalizedDisplayNames.resx` needed? I prefer this approach over testing the implementation details of the string localizer.

dougbu commented on test/Microsoft.AspNetCore.Mvc.DataAnnotations.Test/Internal/DataAnnotationsMetadataProviderTest.cs:
```diff
@@ -727,6 +772,91 @@ public void CreateDisplayMetadata_IsFlagsEnum_ReflectsModelType(Type type, bool
+        }
+
+        [Fact]
+        public void CreateDisplayMetadata_EnumGroupedDisplayNamesAndValues_ResourceTypeLocalizes()
```
Add similar dual-locale tests for `Description`, `DisplayName`, and `Placeholder`. Now that you've fixed `TestResources`, should work w/ either an additional `.resx` for another locale or more properties like `Type_Three`. Up to you.

I'll admit these additional tests aren't about a specific bug. But the test gap you identified exists there as well.

dougbu commented:
> This PR addresses a bug

What bug number? Do we have anything tracking the bug or either of the gaps mentioned in the description?

Not sure about doing the three separate pieces in one PR. I'm not confident the added tests are covering each bit sufficiently.

dougbu commented:
:watch:

ryanbrandenburg commented:
:up::date: I started to split this up into three separate PR's but they're connected enough that it was going to take more time than I though was really prudent to spend on it. Are there any specific test case you feel are missing?

dougbu commented on src/Microsoft.AspNetCore.Mvc.Abstractions/ModelBinding/EnumGroupAndName.cs:
```diff
@@ -10,8 +14,12 @@ namespace Microsoft.AspNetCore.Mvc.ModelBinding
+
         /// <summary>
-        /// Initializes a new instance of the EnumGroupAndName structure.
+        /// Initializes a new instance of the EnumGroupAndName structure. Should not be used if localization is in use.
```
What does "in use" mean? This method should not be used in any site where users may choose multiple locales.

dougbu commented on src/Microsoft.AspNetCore.Mvc.Abstractions/ModelBinding/EnumGroupAndName.cs:
```diff
@@ -39,6 +77,59 @@ public EnumGroupAndName(string group, string name)
+            }
+        }
+
+        // Equals and GetHashCode must be overloaded to accomidate the _nameFunc
```
accomidate -> accommodate

dougbu commented on src/Microsoft.AspNetCore.Mvc.Abstractions/ModelBinding/EnumGroupAndName.cs:
```diff
@@ -39,6 +77,59 @@ public EnumGroupAndName(string group, string name)
+            }
+            else
+            {
+                var display = _fieldInfo.GetCustomAttribute<DisplayAttribute>(inherit: false);
```
This call is expensive and not something we need to do repeatedly. Add another field to store the `DisplayAttribute` and set `_name` to `_fieldInfo.Name` so that this is done at most once even when there's no such attribute.

dougbu replied:
Guess you'll need to do `_fieldInfo = null` as well unless you change the first condition to `_name != null`.

dougbu commented on src/Microsoft.AspNetCore.Mvc.Abstractions/ModelBinding/EnumGroupAndName.cs:
```diff
@@ -39,6 +77,59 @@ public EnumGroupAndName(string group, string name)
+                    {
+                        name = _stringLocalizer[name];
+                    }
+                    return name ?? _fieldInfo.Name;
```
nit: Add a blank line above this.

dougbu commented on src/Microsoft.AspNetCore.Mvc.Abstractions/project.json:
```diff
@@ -33,17 +33,23 @@
       "version": "1.1.0-*",
       "type": "build"
     },
+    "Microsoft.Extensions.Localization.Abstractions": "1.1.0-*",
```
This new dependency may have implications w.r.t. binary compatibility when using just `Microsoft.AspNetCore.Mvc.Core`. Thoughts @rynowak?

dougbu replied:
@rynowak and I discussed this offline and the additional dependencies _are_ a concern for him. We also discussed alternatives and the best we can see is the `Func<string>` approach you were using around 7 iterations ago. That's still an ugly approach but it's better than introducing these dependencies. Sorry.

dougbu commented on src/Microsoft.AspNetCore.Mvc.ViewFeatures/Internal/ExpressionMetadataProvider.cs:
```diff
@@ -2,6 +2,7 @@
 // Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
 
 using System;
+using System.ComponentModel.DataAnnotations;
```
Huh?

dougbu commented on src/Microsoft.AspNetCore.Mvc.ViewFeatures/ViewFeatures/HtmlHelper.cs:
```diff
@@ -522,6 +522,24 @@ public Task RenderPartialAsync(string partialViewName, object model, ViewDataDic
+                var value = (modelExplorer.Model as Enum).ToString("d");
+
+                var enumGrouped = modelExplorer.Metadata.EnumGroupedDisplayNamesAndValues;
+                if (enumGrouped != null &&
```
Since `IsEnum` you can `Debug.Assert(enumGrouped != null)`.

dougbu commented on src/Microsoft.AspNetCore.Mvc.ViewFeatures/ViewFeatures/HtmlHelper.cs:
```diff
@@ -522,6 +522,24 @@ public Task RenderPartialAsync(string partialViewName, object model, ViewDataDic
+
+                var enumGrouped = modelExplorer.Metadata.EnumGroupedDisplayNamesAndValues;
+                if (enumGrouped != null &&
+                    enumGrouped.Count() > 0)
```
Not sure the corner case of an empty `enum` (i.e. one that has no defined values) is worth a condition at all.

dougbu commented on src/Microsoft.AspNetCore.Mvc.ViewFeatures/ViewFeatures/HtmlHelper.cs:
```diff
@@ -522,6 +522,24 @@ public Task RenderPartialAsync(string partialViewName, object model, ViewDataDic
+                    {
+                        if (kvp.Value == value)
+                        {
+                            modelExplorer = modelExplorer.GetExplorerForModel(kvp.Key.Name);
```
Deserves a comment that this will create a `ModelExplorer` with the same `Metadata` but the `Model` will be the overridden `string` representation of the `enum` value.

dougbu replied:
Also `break` after this line since multiple `enum` constants can have the same `Value`.

dougbu commented on test/Microsoft.AspNetCore.Mvc.DataAnnotations.Test/Internal/DataAnnotationsMetadataProviderTest.cs:
```diff
@@ -589,6 +601,49 @@ public void CreateDisplayMetadata_IsFlagsEnum_ReflectsModelType(Type type, bool
         }
 
+        [Fact]
+        public void CreateDisplayMetadata_DisplayName_Localized()
```
Suggest `CreateDisplayMetadata_DisplayName_LocalizeWithStringLocalizer()` to clearly differentiate from tests using `ResourceType` approach.

dougbu commented on test/Microsoft.AspNetCore.Mvc.DataAnnotations.Test/Internal/DataAnnotationsMetadataProviderTest.cs:
```diff
@@ -589,6 +601,49 @@ public void CreateDisplayMetadata_IsFlagsEnum_ReflectsModelType(Type type, bool
+                expectedKeyValuePairs?.OrderBy(item => item.Key.Group, StringComparer.Ordinal)
+                .ThenBy(item => item.Key.Name, StringComparer.Ordinal),
+                context.DisplayMetadata.EnumGroupedDisplayNamesAndValues?.OrderBy(item => item.Key.Group, StringComparer.Ordinal)
+                .ThenBy(item => item.Key.Name, StringComparer.Ordinal));
```
Move the `OrderBy()` calls onto their own lines and indent a bit to make this easier to read.

dougbu replied:
Also, why `null` check `expectedKeyValuePairs`? Remove `null` propagation and at most `Assert.NotNull(context.DisplayMetadata.EnumGroupedDisplayNamesAndValues)`.

dougbu commented on test/Microsoft.AspNetCore.Mvc.DataAnnotations.Test/Internal/DataAnnotationsMetadataProviderTest.cs:
```diff
@@ -589,6 +601,49 @@ public void CreateDisplayMetadata_IsFlagsEnum_ReflectsModelType(Type type, bool
+            provider.CreateDisplayMetadata(context);
+
+            // Assert
+            var expectedKeyValuePairs = new List<KeyValuePair<EnumGroupAndName, string>>
```
This statement should be in `// Arrange`.

dougbu commented on test/Microsoft.AspNetCore.Mvc.DataAnnotations.Test/Internal/DataAnnotationsMetadataProviderTest.cs:
```diff
@@ -727,6 +782,118 @@ public void CreateDisplayMetadata_IsFlagsEnum_ReflectsModelType(Type type, bool
                 .ThenBy(item => item.Key.Name, StringComparer.Ordinal));
         }
 
+        private DataAnnotationsMetadataProvider CreateLocalizingProvider()
```
Please move this down to the bottom of the class, perhaps just above the nested classes. And rename it to mention `IStringLocalizer`. Current name is too broad.

dougbu replied:
Not looking at other new test names 'til you check them throughout this PR. `ResourceType` is a supported way to localize `string`s in the system.

dougbu commented on test/Microsoft.AspNetCore.Mvc.DataAnnotations.Test/Internal/DataAnnotationsMetadataProviderTest.cs:
```diff
@@ -727,6 +782,118 @@ public void CreateDisplayMetadata_IsFlagsEnum_ReflectsModelType(Type type, bool
+                {
+                    if (k.Contains("Loc_Two"))
+                    {
+                        return new LocalizedString(k, $"{k} {CultureInfo.CurrentCulture}");
```
Why isn't this used unconditionally?

ryanbrandenburg replied:
Because we also need to test that we return the DisplayAttribute.Name when there's nothing in the IStringLocalizer.

dougbu replied:
You're testing internal details of the `IStringLocalizer` again. The caller gets a string and deals with it. None of those callers check `ResourceNotFound`

dougbu commented on test/Microsoft.AspNetCore.Mvc.DataAnnotations.Test/Internal/DataAnnotationsMetadataProviderTest.cs:
```diff
@@ -727,6 +782,118 @@ public void CreateDisplayMetadata_IsFlagsEnum_ReflectsModelType(Type type, bool
+            var key = ModelMetadataIdentity.ForType(typeof(EnumWithLocalizedDisplayNames));
+            var attributes = new object[0];
+
+            // Act
```
Move down a couple of lines.

dougbu commented on test/Microsoft.AspNetCore.Mvc.DataAnnotations.Test/Internal/DataAnnotationsMetadataProviderTest.cs:
```diff
@@ -727,6 +782,118 @@ public void CreateDisplayMetadata_IsFlagsEnum_ReflectsModelType(Type type, bool
+            }
+
+            // Assert
+            Assert.NotEqual(frenchEnumDisplay, englishEnumDisplay);
```
Remove.

dougbu commented on test/Microsoft.AspNetCore.Mvc.DataAnnotations.Test/Internal/DataAnnotationsMetadataProviderTest.cs:
```diff
@@ -727,6 +782,118 @@ public void CreateDisplayMetadata_IsFlagsEnum_ReflectsModelType(Type type, bool
+            {
+                frenchEnumDisplay = context.DisplayMetadata.EnumGroupedDisplayNamesAndValues
+                    .Where(kvp => kvp.Value == "2")
+                    .First().Key.Name;
```
If the localizer always localized, could instead save `EnumGroupedDisplayNamesAndValues.Select(kvp => kvp.Key.Name).ToList()`. Then `Assert.All(frenchEnumNames, name => Assert.EndsWith(" fr-FR", name, StringComparison.Ordinal)` down below.

For future reference, don't separate out the `Where()` condition; use `First(kvp => kvp.Value == "2")`.

dougbu commented on test/Microsoft.AspNetCore.Mvc.DataAnnotations.Test/Internal/DataAnnotationsMetadataProviderTest.cs:
```diff
@@ -727,6 +782,118 @@ public void CreateDisplayMetadata_IsFlagsEnum_ReflectsModelType(Type type, bool
+        }
+
+        [Fact]
+        public void CreateDisplayMetadata_EnumGroupedDisplayNamesAndValues_NameLocalizes()
```
This looks like either 3 or 6 individual tests. Figure out a way to make it a `[Theory]`.

dougbu commented on test/Microsoft.AspNetCore.Mvc.ViewFeatures.Test/Rendering/HtmlHelperDisplayExtensionsTest.cs:
```diff
@@ -224,6 +226,75 @@ public void DisplayFor_UsesTemplateNameAndAdditionalViewData()
+                .Setup(s => s["Resx_Created"])
+                .Returns<string>((key) =>
+                {
+                    return new LocalizedString(key, "created from resx");
```
`"created from IStringLocalizer"`

dougbu commented on test/Microsoft.AspNetCore.Mvc.ViewFeatures.Test/Rendering/HtmlHelperDisplayExtensionsTest.cs:
```diff
@@ -356,5 +427,24 @@ private class SomeModel
+
+        public class StatusResource
+        {
+            public static string Type_Faulted { get { return "faulted from type"; } }
```
`"Faulted from ResourceType"`

dougbu commented on test/Microsoft.AspNetCore.Mvc.ViewFeatures.Test/Rendering/HtmlHelperSelectTest.cs:
```diff
@@ -1299,6 +1301,30 @@ public void GetEnumSelectListTEnum_ThrowsWithNonEnum()
+            var result = htmlHelper.GetEnumSelectList<EnumWithDisplayNames>();
+
+            // Assert
+            Assert.Equal("cero en-US", result.Where(s => s.Value == "0").First().Text);
```
Do the same `Assert.All()` thing I mentioned [earlier](https://github.com/aspnet/Mvc/pull/5185/files#r77050742).

dougbu commented:
:watch:

ryanbrandenburg commented:
:up::date: looks like the only unaddressed thing was the new dependency.

dougbu commented on src/Microsoft.AspNetCore.Mvc.Abstractions/ModelBinding/EnumGroupAndName.cs:
```diff
@@ -10,8 +14,14 @@ namespace Microsoft.AspNetCore.Mvc.ModelBinding
         /// <summary>
-        /// Initializes a new instance of the EnumGroupAndName structure.
+        /// Initializes a new instance of the EnumGroupAndName structure. This constructor should not be used in any 
+        /// site where users may select a locale.
```
"users may select a locale" -> "multiple locales may be in use"

ryanbrandenburg replied:
My concern with "multiple locales may be in use" is that it sounds like you're allowing a single user to have multiple simultaneous locals rather than the option between a number of them.

dougbu commented on src/Microsoft.AspNetCore.Mvc.Abstractions/ModelBinding/EnumGroupAndName.cs:
```diff
@@ -39,6 +81,59 @@ public EnumGroupAndName(string group, string name)
+            {
+                if (_displayAttribute != null)
+                {
+                    // Note [Display(Name = "")] is allowed.
```
Looks out of place because the returned value isn't passed to the localizer if it's `string.Empty`. Suggest adding something like `... but we will not attempt to localize and empty name`.

dougbu commented on src/Microsoft.AspNetCore.Mvc.ViewFeatures/ViewFeatures/HtmlHelper.cs:
```diff
@@ -522,6 +523,23 @@ public Task RenderPartialAsync(string partialViewName, object model, ViewDataDic
         {
+            if (modelExplorer.Metadata.IsEnum && modelExplorer.Model != null)
+            {
+                var value = (modelExplorer.Model as Enum).ToString("d");
```
This will hit an NRE if `Model` is `null` which can happen because `ModelExplorer` doesn't enforce that `Model` is compatible with `ModelType` e.g. `string`s are sometimes used. `Model` can also be `null` and compatible w/ `ModelType` i.e. when `ModelType` is a `Nullable<TEnum>`. Suggest saving the `as Model` result in a variable then `null`-checking that variable in the condition just above.

General point is to never dereference a conditional cast without handling `null`.

Think I missed this in the last iteration, sorry.

dougbu commented on test/Microsoft.AspNetCore.Mvc.DataAnnotations.Test/Internal/DataAnnotationsMetadataProviderTest.cs:
```diff
@@ -728,6 +785,96 @@ public void CreateDisplayMetadata_IsFlagsEnum_ReflectsModelType(Type type, bool
+                Assert.All(
+                    context.DisplayMetadata.EnumGroupedDisplayNamesAndValues
+                        .Where(kvp => kvp.Value == "2"),
+                    kvp => Assert.Equal("Loc_Two_Name fr-FR", kvp.Key.Name));
```
This isn't what I suggested. Why are you checking only a single element of the `EnumGroupedDisplayNamesAndValues` collection?

If you must check just one, do what is done elsewhere e.g.
``` c#
var kvp = Assert.Single(context.DisplayMetadata.EnumGroupedDisplayNamesAndValues, k => k.Value == "2");
Assert.Equal("Loc_Two_Name fr-FR", kvp.Key.Name);
```
Even there, note `==` should almost never be used for `string` comparisons and `Assert.Equal()` helpfully includes an overload that accepts `StringComparer.Ordinal`. Be explicit IOW.

ryanbrandenburg replied:
This tests identical functionality to `CreateDisplayMetadata_EnumGroupedDisplayNamesAndValues_NameWithIStringLocalizer` anyway, so I'm just going to delete it unless I'm misunderstanding you and your real problem is with what this test is covering.

dougbu commented on test/Microsoft.AspNetCore.Mvc.DataAnnotations.Test/Internal/DataAnnotationsMetadataProviderTest.cs:
```diff
@@ -728,6 +785,96 @@ public void CreateDisplayMetadata_IsFlagsEnum_ReflectsModelType(Type type, bool
+        public void CreateDisplayMetadata_EnumGroupedDisplayNamesAndValues_NameWithOnlyDisplayAttribute()
+        {
+            // Arrange & Act
+            var enumNameAndGroup = GetLocalizedEnumGrouopedDisplayNamesAndValues();
```
typo in Grouped

dougbu commented on test/Microsoft.AspNetCore.Mvc.DataAnnotations.Test/Internal/DataAnnotationsMetadataProviderTest.cs:
```diff
@@ -878,6 +1063,20 @@ private enum EmptyEnum
         {
         }
 
+        private enum EnumWithLocalizedDisplayNames
```
Why does this `enum` use `Description` or `Prompt`? Only `Name` and `Group` area relevant.

Speaking of which: Think we need to do the same `IStringLocalizer` work for `Group`. Separate issue but worth filing.

ryanbrandenburg replied:
They were leftovers from a previous iteration, cleaned now.

Filed #5217 for Group.

dougbu commented on test/Microsoft.AspNetCore.Mvc.DataAnnotations.Test/Internal/DataAnnotationsMetadataProviderTest.cs:
```diff
@@ -848,6 +995,44 @@ public void CreateValidationDetails_ValidatableObject_AlreadyInContext_Ignores()
+                .Setup(loc => loc[It.IsAny<string>()])
+                .Returns<string>((k =>
+                {
+                    if (k.Contains("Loc_Two") || k.Contains("Type_Three"))
```
Again, remove this conditionality. It means nothing to the code you're testing and complicates things.

ryanbrandenburg replied:
We need it to return the unaltered string (like the localizer would) for the "One" enum in order to ensure that it returns the unaltered Display string. We need it to "localize" for "Type_Three" so that we can ensure we don't actually try to use `IStringLocalizer` on a ResourceType Enum. If we unconditionally alter the string we would be unable to test the case that an enum has a `DisplayAttribute` but doesn't have an entry in the `IStringLocalizer`.

dougbu commented on test/Microsoft.AspNetCore.Mvc.DataAnnotations.Test/Internal/DataAnnotationsMetadataProviderTest.cs:
```diff
@@ -728,6 +785,96 @@ public void CreateDisplayMetadata_IsFlagsEnum_ReflectsModelType(Type type, bool
+        }
+
+        [Fact]
+        public void CreateDisplayMetadata_EnumGroupedDisplayNamesAndValues_NameWithOnlyDisplayAttribute()
```
This is a unit test of `EnumGroupAndName`.

This test also covers an uninteresting scenario. Since only `ModelMetadata` stores instances of this class, `ModelMetadata.EnumGroupedDisplayNamesAndValues` could return a new collection every time it's read and MVC's behaviour would be the same.

dougbu replied:
Same for similar tests below.

ryanbrandenburg replied:
Let's talk about this one in person.

dougbu commented on src/Microsoft.AspNetCore.Mvc.Abstractions/ModelBinding/EnumGroupAndName.cs:
```diff
@@ -39,6 +81,59 @@ public EnumGroupAndName(string group, string name)
+
+            var second = (EnumGroupAndName)obj;
+
+            return string.Equals(Group, second.Group) && string.Equals(Name, second.Name);
```
Be explicit about the `StringComparison` -- always.

dougbu commented on src/Microsoft.AspNetCore.Mvc.Abstractions/ModelBinding/EnumGroupAndName.cs:
```diff
@@ -39,6 +81,59 @@ public EnumGroupAndName(string group, string name)
+            var hashcode = HashCodeCombiner.Start();
+
+            hashcode.Add(Group);
+            hashcode.Add(Name);
```
Instead `Add()` all of the fields. This method must return a single value, regardless of `CurrentCulture`.

ryanbrandenburg replied:
Isn't the same also true for Equals? We won't have access to the fields of the second object within equals, so how should we handle it.

dougbu commented on src/Microsoft.AspNetCore.Mvc.Abstractions/ModelBinding/EnumGroupAndName.cs:
```diff
@@ -28,7 +38,39 @@ public EnumGroupAndName(string group, string name)
+            Group = group;
+            _stringLocalizer = stringLocalizer;
+            _fieldInfo = fieldInfo;
+            _name = null;
```
Only use `_fieldInfo.Name` now. Can store that in `_name` and remove the `_fieldInfo` field.

dougbu commented on test/Microsoft.AspNetCore.Mvc.DataAnnotations.Test/Internal/DataAnnotationsMetadataProviderTest.cs:
```diff
@@ -589,6 +601,51 @@ public void CreateDisplayMetadata_IsFlagsEnum_ReflectsModelType(Type type, bool
+                    .ThenBy(item => item.Key.Name, StringComparer.Ordinal),
+                context.DisplayMetadata.EnumGroupedDisplayNamesAndValues
+                    .OrderBy(item => item.Key.Group, StringComparer.Ordinal)
+                    .ThenBy(item => item.Key.Name, StringComparer.Ordinal));
```
Remove all re-ordering here. `EnumGroupedDisplayNamesAndValues` is an _ordered_ enumeration and that order is important.

dougbu commented on test/Microsoft.AspNetCore.Mvc.DataAnnotations.Test/Internal/DataAnnotationsMetadataProviderTest.cs:
```diff
@@ -728,6 +785,96 @@ public void CreateDisplayMetadata_IsFlagsEnum_ReflectsModelType(Type type, bool
+            // Arrange & Act
+            var enumNameAndGroup = GetLocalizedEnumGrouopedDisplayNamesAndValues();
+
+            var enumOne = enumNameAndGroup.First(e => e.Value == "1");
```
In general, avoid `Linq` in tests because the methods `throw` unreadable `Exceptions` and xUnit.net gives you everything you need. In this case,
``` c#
var enumOne = Assert.Single(enumNameAndGroup, e => e.Value == "1"); // guard
```

dougbu commented on test/Microsoft.AspNetCore.Mvc.ViewFeatures.Test/Rendering/HtmlHelperDisplayExtensionsTest.cs:
```diff
@@ -356,5 +427,24 @@ private class SomeModel
+
+        private enum Status : byte
+        {
+            [Display(Name = "Resx_Created")]
```
Find both display names confusing here. Perhaps it's the fact `.resx` files are used in both `IStringLocalizer` and `ResourceType` approaches and `Type` is much broader than `ResourceType`. These lookup keys should differ from the property names but don't have to be more unique than say `"CreatedKey"` and `"FaultedKey"`.

dougbu commented on test/Microsoft.AspNetCore.Mvc.ViewFeatures.Test/Rendering/HtmlHelperDisplayExtensionsTest.cs:
```diff
@@ -224,6 +226,75 @@ public void DisplayFor_UsesTemplateNameAndAdditionalViewData()
+            var displayResult = helper.DisplayFor(m => m.Status);
+
+            // Assert
+            Assert.Equal("faulted from type", HtmlContentUtilities.HtmlContentToString(displayResult));
```
Does this test pass? The `StatusResource` property returns `"Faulted from ResourceType"`.

dougbu commented on test/Microsoft.AspNetCore.Mvc.ViewFeatures.Test/Rendering/HtmlHelperSelectTest.cs:
```diff
@@ -1299,6 +1301,30 @@ public void GetEnumSelectListTEnum_ThrowsWithNonEnum()
         }
 
         [Fact]
+        public void GetEnumSelectListTEnum_LocalizedEnum()
```
Do not use "Localized" as if it means "using an `IStringLocalizer`". Change test name to describe the specific scenario.

dougbu commented:
:watch: let's sit down and go through your next iteration together.

dougbu commented on src/Microsoft.AspNetCore.Mvc.Abstractions/ModelBinding/EnumGroupAndName.cs:
```diff
@@ -39,6 +77,53 @@ public EnumGroupAndName(string group, string name)
+
+            var second = (EnumGroupAndName)obj;
+
+            return string.Equals(Group, second.Group, StringComparison.Ordinal) && string.Equals(Name, second.Name, StringComparison.Ordinal);
```
Wrap line

dougbu replied:
For consistency and because it's probably better to avoid `IStringLocalizer`'s allocations, check equality of the fields.

dougbu commented on src/Microsoft.AspNetCore.Mvc.ViewFeatures/ViewFeatures/HtmlHelper.cs:
```diff
@@ -522,6 +523,23 @@ public Task RenderPartialAsync(string partialViewName, object model, ViewDataDic
             object additionalViewData)
         {
+            var modelEnum = modelExplorer.Model as Enum;
+            if (modelExplorer.Metadata.IsEnum && modelExplorer.Model != null && modelEnum != null)
```
Don't need `&& modelExplorer.Model != null`

dougbu commented on test/Microsoft.AspNetCore.Mvc.DataAnnotations.Test/Internal/DataAnnotationsMetadataProviderTest.cs:
```diff
@@ -728,6 +779,46 @@ public void CreateDisplayMetadata_IsFlagsEnum_ReflectsModelType(Type type, bool
         }
 
         [Fact]
+        public void CreateDisplayMetadata_EnumGroupedDisplayNamesAndValues_NameWithIStringLocalizer()
```
Suggest creating a small `EnumGroupAndNameTest` class and moving a few of these tests there. No need for duplication here because other tests demonstrate the `EnumGroupAndName` `struct`s are created correctly in general (not much logic left in the provider).

dougbu replied:
Cover 4 paths through `EnumGroupAndName.GetDisplayName()` to make sure we don't do our usual `GetName()` -> `Name` regression in the future.

dougbu commented on test/Microsoft.AspNetCore.Mvc.DataAnnotations.Test/Internal/DataAnnotationsMetadataProviderTest.cs:
```diff
@@ -728,6 +779,46 @@ public void CreateDisplayMetadata_IsFlagsEnum_ReflectsModelType(Type type, bool
+        public void CreateDisplayMetadata_EnumGroupedDisplayNamesAndValues_NameWithIStringLocalizer()
+        {
+            // Arrange & Act
+            var enumNameAndGroup = GetLocalizedEnumGroupedDisplayNamesAndValues();
```
`enumGroupAndName`

dougbu commented:
:watch: mainly for `EnumGroupAndNameTest` refactoring

ryanbrandenburg commented on test/Microsoft.AspNetCore.Mvc.DataAnnotations.Test/Internal/DataAnnotationsMetadataProviderTest.cs:
```diff
@@ -728,6 +779,46 @@ public void CreateDisplayMetadata_IsFlagsEnum_ReflectsModelType(Type type, bool
         }
```
`CreateDisplayMetadata_EnumGroupedDisplayNamesAndValues_ReflectsModelType` fails with our new EnumGroupAndName.Equals because all the EnumGroupAndName's  in the MemberData don't use `FieldInfo` (and thus _displayattribute). Would it be enough for this test to check that the Name and group values on `EnumGroupedDisplayNamesAndValues` are what they should be? Because changing the MemberData to use the FieldInfo's would be basically testing that it equals itself, correct?

dougbu replied:
Yeah, best to check `Name` and `Group` in that test. Might want to create a `EnumGroupAndNameComparer` if that needs to be done in more than one place.

dougbu replied:
:see_no_evil: this comment thread is no longer relevant.

ryanbrandenburg commented:
:up::date: Good idea on the comparer, made it much cleaner than the silly foreach zip I had started trying to do.

ryanbrandenburg commented:
:up::date:

dougbu commented on src/Microsoft.AspNetCore.Mvc.Abstractions/ModelBinding/EnumGroupAndName.cs:
```diff
@@ -28,7 +32,30 @@ public EnumGroupAndName(string group, string name)
+        /// Initializes a new instance of the EnumGroupAndName structure.
+        /// </summary>
+        /// <param name="group">The group name.</param>
+        /// <param name="name">A func which will return the name</param>
```
End w/ a period.

dougbu commented on src/Microsoft.AspNetCore.Mvc.Abstractions/ModelBinding/EnumGroupAndName.cs:
```diff
@@ -39,6 +66,36 @@ public EnumGroupAndName(string group, string name)
+            var second = (EnumGroupAndName)obj;
+
+            return string.Equals(Group, second.Group, StringComparison.Ordinal)
+                && _name == second._name;
```
`Func` equality is not a straightforward thing. But is `EnumGroupAndName` equality important anywhere? I know we create `IList<KeyValuePair<EnumGroupAndName, string>>` collections but does any product code require a meaningful `Equals()` e.g. does the product code use any `IDictionary<EnumGroupAndName, string>` instances?

I'd rather `EnumGroupAndName` avoid overriding `Equals()` and `GetHashCode()` if possible. Fine to create a test-only `EnumGroupAndNameComparer` which uses the `Group` and `Name` properties (perhaps in `CultureInfo.InvariantCulture` to be consistent).

ryanbrandenburg replied:
Avoiding overriding `Equals` and `GetHashCode` would be my preference, but I thought you had said we needed to because the default implementation would be wrong. I don't know of anywhere that will use it so I'll happily remove these two.

dougbu commented on src/Microsoft.AspNetCore.Mvc.Abstractions/ModelBinding/EnumGroupAndName.cs:
```diff
@@ -2,6 +2,7 @@
 // Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
 
 using System;
+using Microsoft.Extensions.Internal;
```
Won't need this if you can remove the `Equals()` and `GetHashCode()` overrides.

dougbu commented on src/Microsoft.AspNetCore.Mvc.DataAnnotations/Internal/DataAnnotationsMetadataProvider.cs:
```diff
@@ -291,38 +293,39 @@ public void CreateValidationMetadata(ValidationMetadataProviderContext context)
             }
         }
 
-        // Return non-empty name specified in a [Display] attribute for a field, if any; field.Name otherwise.
```
Return these methods to their previous order. No longer sure what changed here.

dougbu replied:
thx

dougbu commented on src/Microsoft.AspNetCore.Mvc.DataAnnotations/Internal/DataAnnotationsMetadataProvider.cs:
```diff
@@ -153,15 +153,17 @@ public void CreateDisplayMetadata(DisplayMetadataProviderContext context)
-                        new EnumGroupAndName(groupName, displayName),
+                        new EnumGroupAndName(
+                            groupName,
+                            () => { return GetDisplayName(field, enumLocalizer); }),
```
Minor but ick: Don't be so verbose when `() => GetDisplayName(field, enumLocalizer)` will do.

dougbu commented on test/Microsoft.AspNetCore.Mvc.DataAnnotations.Test/Internal/DataAnnotationsMetadataProviderTest.cs:
```diff
@@ -848,6 +987,58 @@ public void CreateValidationDetails_ValidatableObject_AlreadyInContext_Ignores()
+                useStringLocalizer ? stringLocalizerFactory.Object : null);
+        }
+
+        private class KeyValuePairEnumGroupAndName : IEqualityComparer<KeyValuePair<EnumGroupAndName, string>>
```
Have a look at other test-only comparers. They all have names like `EnumGroupAndNameComparer` and `static` `Instance` or sometimes `Comparer` properties.

dougbu commented on test/Microsoft.AspNetCore.Mvc.DataAnnotations.Test/Internal/DataAnnotationsMetadataProviderTest.cs:
```diff
@@ -848,6 +987,58 @@ public void CreateValidationDetails_ValidatableObject_AlreadyInContext_Ignores()
+            public bool Equals(KeyValuePair<EnumGroupAndName, string> x, KeyValuePair<EnumGroupAndName, string> y)
+            {
+                return x.Key.Name.Equals(y.Key.Name, StringComparison.Ordinal)
+                    && x.Key.Group.Equals(y.Key.Group, StringComparison.Ordinal);
```
Do this w/in a `using (new CultureReplacer(string.Empty, string.Empty))` to play fair. Same for `GetHashCode()`.

dougbu commented on test/Microsoft.AspNetCore.Mvc.DataAnnotations.Test/Internal/DataAnnotationsMetadataProviderTest.cs:
```diff
@@ -874,6 +1065,16 @@ private class ClassWithProperties
+            Two = 2,
+            [Display(
+                Name = "Type_Three_Name",
+                ResourceType = typeof(TestResources))]
```
Can this fit on one line?

dougbu commented on src/Microsoft.AspNetCore.Mvc.DataAnnotations/Internal/DataAnnotationsMetadataProvider.cs:
```diff
@@ -291,38 +293,39 @@ public void CreateValidationMetadata(ValidationMetadataProviderContext context)
 
-        // Return non-empty group specified in a [Display] attribute for a field, if any; string.Empty otherwise.
-        private static string GetDisplayGroup(FieldInfo field)
+        private static string GetDisplayName(FieldInfo fieldInfo, IStringLocalizer stringLocalizer)
```
For the future: If we see this method on the containing `Func` show up in perf profiles, might be worthwhile for this method to return `Func<string>` instead. Most of the decision tree is known when the `Func` is created. Even `string.IsNullOrEmpty(name)` won't vary when `displayAttribute.ResourceType == null`.

dougbu commented on test/Microsoft.AspNetCore.Mvc.DataAnnotations.Test/Internal/DataAnnotationsMetadataProviderTest.cs:
```diff
@@ -720,11 +776,94 @@ public void CreateDisplayMetadata_IsFlagsEnum_ReflectsModelType(Type type, bool
+                .ThenBy(item => item.Key.Name, StringComparer.Ordinal);
+            var actualOrdered = context.DisplayMetadata.EnumGroupedDisplayNamesAndValues?
+                .OrderBy(item => item.Key.Group, StringComparer.Ordinal)
+                .ThenBy(item => item.Key.Name, StringComparer.Ordinal);
```
As I said in a previous iteration, we should never reorder `EnumGroupedDisplayNamesAndValues` in tests. The order of that collection is important and should always be tested.

dougbu replied:
Put another way, fix the data set if `expectedKeyValuePairs` order is wrong. It'll always be wrong and the comment just above is also wrong.

dougbu commented on test/Microsoft.AspNetCore.Mvc.DataAnnotations.Test/project.json:
```diff
@@ -6,6 +6,7 @@
   "dependencies": {
     "dotnet-test-xunit": "2.2.0-*",
     "Microsoft.AspNetCore.Mvc": "1.1.0-*",
+    "Microsoft.AspNetCore.Mvc.DataAnnotations": "1.1.0-*",
```
Remove. If this is needed along w/ the Mvc meta-package, something is wrong.

dougbu commented:
:watch: for some cleanup

ryanbrandenburg commented:
:up::date:

dougbu commented on test/Microsoft.AspNetCore.Mvc.DataAnnotations.Test/Internal/TestResources.cs:
```diff
@@ -10,11 +10,15 @@ namespace Microsoft.AspNetCore.Mvc.ModelBinding
-        public static string DisplayAttribute_Description { get; } = Resources.DisplayAttribute_Description;
+        public static string Type_Three_Name => "type three name " + CultureInfo.CurrentCulture;
+        public static string Type_Three_Description => "type three description " + CultureInfo.CurrentCulture;
+        public static string Type_Three_Prompt => "type three prompt " + CultureInfo.CurrentCulture;
```
Please double-check that `[ReplaceCulture]` or `using (new CultureReplacer(...))` is used in all tests referencing the `CurrentCulture` properties from this class.

dougbu replied:
Hmm, didn't notice `Type_Three_Description` and `Type_Three_Prompt` aren't referenced. Remove those two.

dougbu commented on test/Microsoft.AspNetCore.Mvc.ViewFeatures.Test/Rendering/HtmlHelperSelectTest.cs:
```diff
@@ -1299,6 +1301,31 @@ public void GetEnumSelectListTEnum_ThrowsWithNonEnum()
+
+            // Assert
+            var zeroSelect = Assert.Single(result, s => s.Value.Equals("0", StringComparison.Ordinal));
+            Assert.Equal("cero en-US", zeroSelect.Text);
```
This will fail w/ every culture except `en-US`. Add `[ReplaceCulture]` to the test.

dougbu replied:
thx

dougbu commented on src/Microsoft.AspNetCore.Mvc.Abstractions/ModelBinding/EnumGroupAndName.cs:
```diff
@@ -28,7 +31,30 @@ public EnumGroupAndName(string group, string name)
+        /// Initializes a new instance of the EnumGroupAndName structure.
+        /// </summary>
+        /// <param name="group">The group name.</param>
+        /// <param name="name">A func which will return the name.</param>
```
Should have said: `func` -> `<see cref="Func{string}"/>`. Similarly `EnumGroupAndName` -> `<see cref="EnumGroupAndName"/>` above (in the original constructor overload as well).

dougbu commented:
:watch: for the last couple of wrinkles to be ironed out.

ryanbrandenburg commented:
:up::date:

dougbu commented:
:shipit: w/ [one removal](https://github.com/aspnet/Mvc/pull/5185#discussion-diff-77874996) and a [better commit description](https://github.com/aspnet/Mvc/pull/5185#issuecomment-242473921). Also mention the relevant [bugs](https://github.com/aspnet/Mvc/pull/5185#issuecomment-242490114).

ryanbrandenburg commented:
@dougbu are you talking about the actual commit message or about the Pull Request description? Because I already added the # for the 3 bugs and removed (to my eyes) the implication that it's only about localization on the Pull Request description.
